ironfish/src/blockchain/nullifierSet/nullifierSet.ts

```diff
@@ -0,0 +1,132 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { Assert } from '../../assert'
+import { Block, Transaction } from '../../primitives'
+import { Nullifier } from '../../primitives/nullifier'
+import {
+  BUFFER_ENCODING,
+  DatabaseSchema,
+  IDatabase,
+  IDatabaseStore,
+  IDatabaseTransaction,
+  StringEncoding,
+  U32_ENCODING,
+} from '../../storage'
+import { NullifierInfo, NullifierLocationEncoding } from './encoding'
+
+interface CounterEntry<T extends string> extends DatabaseSchema {
+  key: T
+  value: number
+}
+
+interface NullifiersSchema extends DatabaseSchema {
+  key: Nullifier
+  value: NullifierInfo
+}
+
+export class NullifierSet {
+  readonly db: IDatabase
+
+  // Keep track of number of nullifiers in the set
+  private readonly counter: IDatabaseStore<CounterEntry<'Size'>>
+
+  // Nullifier -> BlockHash, TransactionHash
+  private readonly nullifiers: IDatabaseStore<NullifiersSchema>
+
+  constructor(options: { db: IDatabase; storeName: string }) {
+    this.db = options.db
+
+    this.counter = this.db.addStore({
+      name: `${options.storeName}c`,
+      keyEncoding: new StringEncoding<'Size'>(),
+      valueEncoding: U32_ENCODING,
+    })
+
+    // Nullifier -> BlockHash, TransactionHash
+    this.nullifiers = this.db.addStore({
+      name: options.storeName,
+      keyEncoding: BUFFER_ENCODING,
+      valueEncoding: new NullifierLocationEncoding(),
+    })
+  }
+
+  size(tx?: IDatabaseTransaction): Promise<number> {
+    return this.db.withTransaction(tx, async (tx) => {
+      const size = await this.counter.get('Size', tx)
+      return size === undefined ? 0 : size
+    })
+  }
+
+  contains(nullifier: Nullifier, tx?: IDatabaseTransaction): Promise<boolean> {
+    return this.db.withTransaction(tx, async (tx) => {
+      return await this.nullifiers.has(nullifier, tx)
+    })
+  }
+
+  contained(nullifier: Nullifier, size: number, tx?: IDatabaseTransaction): Promise<boolean> {
```

Comment:
Do we need this in a set? It seems like we should delete this function now right?

Reply:
This will be useful for the case where we want to check whether a transaction is a double spend with a given confirmation range. In gossip code/mempool for example if we want to reject double spends that are older than 10 block confirmation range we can get the size at that block and call contained with it

Reply:
Discussed this and for this specific use case we could just keep a cache of nullifiers in previous 10 blocks for example the code would look like
```javascript
function isDoubleSpend(nullifier) {
  if(chain.nullifiers.contains(nullifier)) {
    return !last10BlocksCache.contains(nullifier)
  }
  return false
}
```